### ISSUE TITLE
error for "do undef" references "require" instead

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14916]
Migrated from [rt.perl.org#126084](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=126084) (status was 'resolved')

Searchable as RT126084$

**Comment by p5pRT** [comment]
### From @rjbs

#### Created by @rjbs



&nbsp;  perl \-e 'do undef'

Since 5\.22\.0\, this reports "Missing or undefined argument to require"\,
an improvement on the previous "Null filename used"\.\.\. when it's for
require\.

"do undef" and "do\(\)" should say "\.\.\.to do"\, not "\.\.\.to require"

\(It's too bad we don't traditionally quote the builtins in these messages\, as
"to do" is somewhat ambiguous if you don't know what's up\.  Fortunately\, the
line number should clarify\.\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.22.0:

Configured by rjbs at Mon Jun  1 12:58:35 EDT 2015.

Summary of my perl5 (revision 5 version 22 subversion 0) configuration:
   
  Platform:
    osname=darwin, osvers=14.3.0, archname=darwin-2level
    uname='darwin jubjub.local 14.3.0 darwin kernel version 14.3.0: mon mar 23 11:59:05 pdt 2015; root:xnu-2782.20.48~5release_x86_64 x86_64 '
    config_args='-Dprefix=/Users/rjbs/.plenv/versions/22.0 -de -Dusedevel -A'eval:scriptdir=/Users/rjbs/.plenv/versions/22.0/bin''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -I/opt/local/include',
    optimize='-O3',
    cppflags='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -I/opt/local/include'
    ccversion='', gccversion='4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector-strong -L/usr/local/lib -L/opt/local/lib'
    libpth=/usr/local/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/6.1.0/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib /usr/lib /opt/local/lib
    libs=-lpthread -lgdbm -ldbm -ldl -lm -lutil -lc
    perllibs=-lpthread -ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/usr/local/lib -L/opt/local/lib -fstack-protector-strong'



@INC for perl 5.22.0:
    /Users/rjbs/.plenv/versions/22.0/lib/perl5/site_perl/5.22.0/darwin-2level
    /Users/rjbs/.plenv/versions/22.0/lib/perl5/site_perl/5.22.0
    /Users/rjbs/.plenv/versions/22.0/lib/perl5/5.22.0/darwin-2level
    /Users/rjbs/.plenv/versions/22.0/lib/perl5/5.22.0
    .


Environment for perl 5.22.0:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/rjbs
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/Users/rjbs/.plenv/versions/22.0/bin:/Users/rjbs/.plenv/libexec:/Users/rjbs/.plenv/plugins/perl-build/bin:/Users/rjbs/bin:/Users/rjbs/.rbenv/shims:/Users/rjbs/.rbenv/bin:/Users/rjbs/.rakudobrew/bin:/Users/rjbs/.plenv/shims:/Users/rjbs/.plenv/bin:/opt/local/bin:/opt/local/sbin:/usr/local/texlive/2015/bin/x86_64-darwin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/opt/X11/bin:/Users/rjbs/code/hla
    PERLDOC=-otext
    PERL_AUTOINSTALL=--skipdeps
    PERL_BADLANG (unset)
    PERL_MAILERS=sendmail:/Users/rjbs/bin/sendmail
    SHELL=/opt/local/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Thu Sep 17 06&#8203;:37&#8203;:08 2015\, rjbs wrote&#8203;:
> This is a bug report for perl from rjbs@&#8203;cpan\.org\,
> generated with the help of perlbug 1\.40 running under perl 5\.22\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> perl \-e 'do undef'
> 
> Since 5\.22\.0\, this reports "Missing or undefined argument to require"\,
> an improvement on the previous "Null filename used"\.\.\. when it's for
> require\.
> 
> "do undef" and "do\(\)" should say "\.\.\.to do"\, not "\.\.\.to require"
> 

These warnings are defined in PP\(pp\_require\) in pp\_ctl\.c\.  There are several other points where it is indicated that the dofile opcode is implemented via pp\_require\.  So\, at a high level\, we would probably need a PP\(pp\_do\) in pp\_ctl\.c\.  But pp\_require is several hundreds line of code long; I'm guessing we wouldn't need all of it for a pp\_do\. 





> \(It's too bad we don't traditionally quote the builtins in these
> messages\, as
> "to do" is somewhat ambiguous if you don't know what's up\.
> Fortunately\, the
> line number should clarify\.\)
> 
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=medium
> \-\-\-
> Site configuration information for perl 5\.22\.0&#8203;:
> 
> Configured by rjbs at Mon Jun  1 12&#8203;:58&#8203;:35 EDT 2015\.
> 
> Summary of my perl5 \(revision 5 version 22 subversion 0\)
> configuration&#8203;:
> 
> Platform&#8203;:
>   osname=darwin\, osvers=14\.3\.0\, archname=darwin\-2level
>   uname='darwin jubjub\.local 14\.3\.0 darwin kernel version 14\.3\.0&#8203;: mon
> mar 23 11&#8203;:59&#8203;:05 pdt 2015; root&#8203;:xnu\-2782\.20\.48~5release\_x86\_64 x86\_64 '
>   config\_args='\-Dprefix=/Users/rjbs/\.plenv/versions/22\.0 \-de
> \-Dusedevel \-A'eval&#8203;:scriptdir=/Users/rjbs/\.plenv/versions/22\.0/bin''
>   hint=recommended\, useposix=true\, d\_sigaction=define
>   useithreads=undef\, usemultiplicity=undef
>   use64bitint=define\, use64bitall=define\, uselongdouble=undef
>   usemymalloc=n\, bincompat5005=undef
> Compiler&#8203;:
>   cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing
> \-pipe \-fstack\-protector\-strong \-I/usr/local/include
> \-I/opt/local/include'\,
>   optimize='\-O3'\,
>   cppflags='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe
> \-fstack\-protector\-strong \-I/usr/local/include \-I/opt/local/include'
>   ccversion=''\, gccversion='4\.2\.1 Compatible Apple LLVM 6\.1\.0 \(clang\-
> 602\.0\.53\)'\, gccosandvers=''
>   intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678\,
> doublekind=3
>   d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16\,
> longdblkind=3
>   ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
> lseeksize=8
>   alignbytes=8\, prototype=define
> Linker and Libraries&#8203;:
>   ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-fstack\-
> protector\-strong \-L/usr/local/lib \-L/opt/local/lib'
>   libpth=/usr/local/lib
> /Applications/Xcode\.app/Contents/Developer/Toolchains/XcodeDefault\.xctoolchain/usr/bin/\.\./lib/clang/6\.1\.0/lib
> /Applications/Xcode\.app/Contents/Developer/Toolchains/XcodeDefault\.xctoolchain/usr/lib
> /usr/lib /opt/local/lib
>   libs=\-lpthread \-lgdbm \-ldbm \-ldl \-lm \-lutil \-lc
>   perllibs=\-lpthread \-ldl \-lm \-lutil \-lc
>   libc=\, so=dylib\, useshrplib=false\, libperl=libperl\.a
>   gnulibc\_version=''
> Dynamic Linking&#8203;:
>   dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
>   cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup
> \-L/usr/local/lib \-L/opt/local/lib \-fstack\-protector\-strong'
> 
> 
> \-\-\-
> @&#8203;INC for perl 5\.22\.0&#8203;:
>     /Users/rjbs/\.plenv/versions/22\.0/lib/perl5/site\_perl/5\.22\.0/darwin\-
> 2level
>     /Users/rjbs/\.plenv/versions/22\.0/lib/perl5/site\_perl/5\.22\.0
>     /Users/rjbs/\.plenv/versions/22\.0/lib/perl5/5\.22\.0/darwin\-2level
>     /Users/rjbs/\.plenv/versions/22\.0/lib/perl5/5\.22\.0
>     \.
> 
> \-\-\-
> Environment for perl 5\.22\.0&#8203;:
>     DYLD\_LIBRARY\_PATH \(unset\)
>     HOME=/Users/rjbs
>     LANG=en\_US\.UTF\-8
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/Users/rjbs/\.plenv/versions/22\.0/bin&#8203;:/Users/rjbs/\.plenv/libexec&#8203;:/Users/rjbs/\.plenv/plugins/perl\-
> build/bin&#8203;:/Users/rjbs/bin&#8203;:/Users/rjbs/\.rbenv/shims&#8203;:/Users/rjbs/\.rbenv/bin&#8203;:/Users/rjbs/\.rakudobrew/bin&#8203;:/Users/rjbs/\.plenv/shims&#8203;:/Users/rjbs/\.plenv/bin&#8203;:/opt/local/bin&#8203;:/opt/local/sbin&#8203;:/usr/local/texlive/2015/bin/x86\_64\-
> darwin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/opt/X11/bin&#8203;:/Users/rjbs/code/hla
>     PERLDOC=\-otext
>     PERL\_AUTOINSTALL=\-\-skipdeps
>     PERL\_BADLANG \(unset\)
>     PERL\_MAILERS=sendmail&#8203;:/Users/rjbs/bin/sendmail
>     SHELL=/opt/local/bin/zsh


\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Sep 20 15&#8203;:57&#8203;:21 2015\, jkeenan wrote&#8203;:
> These warnings are defined in PP\(pp\_require\) in pp\_ctl\.c\.  There are
> several other points where it is indicated that the dofile opcode is
> implemented via pp\_require\.  So\, at a high level\, we would probably
> need a PP\(pp\_do\) in pp\_ctl\.c\.  But pp\_require is several hundreds line
> of code long; I'm guessing we wouldn't need all of it for a pp\_do\.

You don't need to duplicate pp\_require\, we have a macro to find the description of an op\.

See the attached\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-126084-report-the-correct-op-in-pp_require.patch</summary>

```diff
From 0cf23372b33f16f0d3977c88c901b7f04b3cb21d Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Mon, 21 Sep 2015 09:43:14 +1000
Subject: [perl #126084] report the correct op in pp_require

pp_require implements both require and do "file", but always reported
the error as being for require.
---
 pod/perldiag.pod      | 12 ++++++++----
 pp_ctl.c              |  8 ++++----
 t/lib/croak/pp_ctl    | 20 ++++++++++++++++++++
 t/lib/warnings/pp_ctl | 13 +++++++++++++
 4 files changed, 45 insertions(+), 8 deletions(-)

diff --git a/pod/perldiag.pod b/pod/perldiag.pod
index 6af1245..eeb658a 100644
--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -3397,11 +3397,15 @@ can vary from one line to the next.
 (S syntax) This is an educated guess made in conjunction with the message
 "%s found where operator expected".  Often the missing operator is a comma.
 
-=item Missing or undefined argument to require
+=item Missing or undefined argument to %s
 
-(F) You tried to call require with no argument or with an undefined
-value as an argument.  Require expects either a package name or a
-file-specification as an argument.  See L<perlfunc/require>.
+(F) You tried to call require or do with no argument or with an undefined
+value as an argument.
+
+Require expects either a package name or a file-specification as an
+argument.  See L<perlfunc/require>.
+
+Do expected a file-specification as an argument.  See L<perlfunc/do>.
 
 =item Missing right brace on \%c{} in regex; marked by S<<-- HERE> in m/%s/
 
diff --git a/pp_ctl.c b/pp_ctl.c
index 852ec36..8e05d53 100644
--- a/pp_ctl.c
+++ b/pp_ctl.c
@@ -3686,12 +3686,12 @@ PP(pp_require)
 	RETPUSHYES;
     }
     if (!SvOK(sv))
-        DIE(aTHX_ "Missing or undefined argument to require");
+        DIE(aTHX_ "Missing or undefined argument to %s", OP_DESC(PL_op));
     name = SvPV_nomg_const(sv, len);
     if (!(name && len > 0 && *name))
-        DIE(aTHX_ "Missing or undefined argument to require");
+        DIE(aTHX_ "Missing or undefined argument to %s", OP_DESC(PL_op));
 
-    if (!IS_SAFE_PATHNAME(name, len, "require")) {
+    if (!IS_SAFE_PATHNAME(name, len, OP_DESC(PL_op))) {
         DIE(aTHX_ "Can't locate %s:   %s",
             pv_escape(newSVpvs_flags("",SVs_TEMP),SvPVX(sv),SvCUR(sv),
                       SvCUR(sv)*2,NULL, SvUTF8(sv)?PERL_PV_ESCAPE_UNI:0),
@@ -3899,7 +3899,7 @@ PP(pp_require)
 			dirlen = 0;
 		    }
 
-		    if (!IS_SAFE_SYSCALL(dir, dirlen, "@INC entry", "require"))
+		    if (!IS_SAFE_SYSCALL(dir, dirlen, "@INC entry", OP_DESC(PL_op)))
 			continue;
 #ifdef VMS
 		    if ((unixdir =
diff --git a/t/lib/croak/pp_ctl b/t/lib/croak/pp_ctl
index ee1edba..0ce9166 100644
--- a/t/lib/croak/pp_ctl
+++ b/t/lib/croak/pp_ctl
@@ -16,3 +16,23 @@ use 5.01;
 default{}
 EXPECT
 Can't "default" outside a topicalizer at - line 2.
+########
+# NAME error on require undef
+require undef
+EXPECT
+Missing or undefined argument to require at - line 1.
+########
+# NAME error on require ""
+require ""
+EXPECT
+Missing or undefined argument to require at - line 1.
+########
+# NAME error on do undef
+do undef
+EXPECT
+Missing or undefined argument to do "file" at - line 1.
+########
+# NAME error on do ""
+do ""
+EXPECT
+Missing or undefined argument to do "file" at - line 1.
diff --git a/t/lib/warnings/pp_ctl b/t/lib/warnings/pp_ctl
index 9b3f298..2959fef 100644
--- a/t/lib/warnings/pp_ctl
+++ b/t/lib/warnings/pp_ctl
@@ -251,3 +251,16 @@ EXPECT
 use warnings;
 eval 'use 5.006; use 5.10.0';
 EXPECT
+########
+# NAME \0 in require/do arguments and @INC
+use warnings;
+eval 'require "a\0b";';
+eval 'do "a\0b"';
+push @INC, "a\0b";
+eval 'require "some-unknown-filename"';
+eval 'do "some-unknown-filename"';
+EXPECT
+Invalid \0 character in pathname for require: a\0b at (eval 1) line 1.
+Invalid \0 character in pathname for do "file": a\0b at (eval 2) line 1.
+Invalid \0 character in @INC entry for require: a\0b at (eval 3) line 1.
+Invalid \0 character in @INC entry for do "file": a\0b at (eval 4) line 1.
-- 
2.1.4


```
</details>

**Comment by p5pRT** [comment]
### From @rjbs

\* Tony Cook via RT \<perlbug\-followup@&#8203;perl\.org> \[2015\-09\-20T19&#8203;:44&#8203;:59\]
> See the attached\.

Thanks\!

> \+Require expects either a package name or a file\-specification as an
> \+argument\.  See L\<perlfunc/require>\.
> \+
> \+Do expected a file\-specification as an argument\.  See L\<perlfunc/do>\.

^\-\-\- Verb tense disagreement

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Sep 20 16&#8203;:50&#8203;:17 2015\, perl\.p5p@&#8203;rjbs\.manxome\.org wrote&#8203;:
> \* Tony Cook via RT \<perlbug\-followup@&#8203;perl\.org> \[2015\-09\-20T19&#8203;:44&#8203;:59\]
> > See the attached\.
> 
> Thanks\!
> 
> > \+Require expects either a package name or a file\-specification as an
> > \+argument\.  See L\<perlfunc/require>\.
> > \+
> > \+Do expected a file\-specification as an argument\.  See L\<perlfunc/do>\.
> 
> ^\-\-\- Verb tense disagreement

Oops\, updated patch\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-126084-report-the-correct-op-in-pp_require.patch</summary>

```diff
From e6c2a82191fb03d6a76ba7e238cdccae1c859af2 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Mon, 21 Sep 2015 10:11:39 +1000
Subject: [perl #126084] report the correct op in pp_require

pp_require implements both require and do "file", but always reported
the error as being for require.
---
 pod/perldiag.pod      | 12 ++++++++----
 pp_ctl.c              |  8 ++++----
 t/lib/croak/pp_ctl    | 20 ++++++++++++++++++++
 t/lib/warnings/pp_ctl | 13 +++++++++++++
 4 files changed, 45 insertions(+), 8 deletions(-)

diff --git a/pod/perldiag.pod b/pod/perldiag.pod
index 6af1245..2e41502 100644
--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -3397,11 +3397,15 @@ can vary from one line to the next.
 (S syntax) This is an educated guess made in conjunction with the message
 "%s found where operator expected".  Often the missing operator is a comma.
 
-=item Missing or undefined argument to require
+=item Missing or undefined argument to %s
 
-(F) You tried to call require with no argument or with an undefined
-value as an argument.  Require expects either a package name or a
-file-specification as an argument.  See L<perlfunc/require>.
+(F) You tried to call require or do with no argument or with an undefined
+value as an argument.
+
+Require expects either a package name or a file-specification as an
+argument.  See L<perlfunc/require>.
+
+Do expects a file-specification as an argument.  See L<perlfunc/do>.
 
 =item Missing right brace on \%c{} in regex; marked by S<<-- HERE> in m/%s/
 
diff --git a/pp_ctl.c b/pp_ctl.c
index 852ec36..8e05d53 100644
--- a/pp_ctl.c
+++ b/pp_ctl.c
@@ -3686,12 +3686,12 @@ PP(pp_require)
 	RETPUSHYES;
     }
     if (!SvOK(sv))
-        DIE(aTHX_ "Missing or undefined argument to require");
+        DIE(aTHX_ "Missing or undefined argument to %s", OP_DESC(PL_op));
     name = SvPV_nomg_const(sv, len);
     if (!(name && len > 0 && *name))
-        DIE(aTHX_ "Missing or undefined argument to require");
+        DIE(aTHX_ "Missing or undefined argument to %s", OP_DESC(PL_op));
 
-    if (!IS_SAFE_PATHNAME(name, len, "require")) {
+    if (!IS_SAFE_PATHNAME(name, len, OP_DESC(PL_op))) {
         DIE(aTHX_ "Can't locate %s:   %s",
             pv_escape(newSVpvs_flags("",SVs_TEMP),SvPVX(sv),SvCUR(sv),
                       SvCUR(sv)*2,NULL, SvUTF8(sv)?PERL_PV_ESCAPE_UNI:0),
@@ -3899,7 +3899,7 @@ PP(pp_require)
 			dirlen = 0;
 		    }
 
-		    if (!IS_SAFE_SYSCALL(dir, dirlen, "@INC entry", "require"))
+		    if (!IS_SAFE_SYSCALL(dir, dirlen, "@INC entry", OP_DESC(PL_op)))
 			continue;
 #ifdef VMS
 		    if ((unixdir =
diff --git a/t/lib/croak/pp_ctl b/t/lib/croak/pp_ctl
index ee1edba..0ce9166 100644
--- a/t/lib/croak/pp_ctl
+++ b/t/lib/croak/pp_ctl
@@ -16,3 +16,23 @@ use 5.01;
 default{}
 EXPECT
 Can't "default" outside a topicalizer at - line 2.
+########
+# NAME error on require undef
+require undef
+EXPECT
+Missing or undefined argument to require at - line 1.
+########
+# NAME error on require ""
+require ""
+EXPECT
+Missing or undefined argument to require at - line 1.
+########
+# NAME error on do undef
+do undef
+EXPECT
+Missing or undefined argument to do "file" at - line 1.
+########
+# NAME error on do ""
+do ""
+EXPECT
+Missing or undefined argument to do "file" at - line 1.
diff --git a/t/lib/warnings/pp_ctl b/t/lib/warnings/pp_ctl
index 9b3f298..2959fef 100644
--- a/t/lib/warnings/pp_ctl
+++ b/t/lib/warnings/pp_ctl
@@ -251,3 +251,16 @@ EXPECT
 use warnings;
 eval 'use 5.006; use 5.10.0';
 EXPECT
+########
+# NAME \0 in require/do arguments and @INC
+use warnings;
+eval 'require "a\0b";';
+eval 'do "a\0b"';
+push @INC, "a\0b";
+eval 'require "some-unknown-filename"';
+eval 'do "some-unknown-filename"';
+EXPECT
+Invalid \0 character in pathname for require: a\0b at (eval 1) line 1.
+Invalid \0 character in pathname for do "file": a\0b at (eval 2) line 1.
+Invalid \0 character in @INC entry for require: a\0b at (eval 3) line 1.
+Invalid \0 character in @INC entry for do "file": a\0b at (eval 4) line 1.
-- 
2.1.4


```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Sep 20 17&#8203;:13&#8203;:19 2015\, tonyc wrote&#8203;:

> Oops\, updated patch\.
> 

\+1\.  Smoking in branch smoke\-me/jkeenan/126084\-dofile\.


\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Sep 21 18&#8203;:28&#8203;:35 2015\, jkeenan wrote&#8203;:
> On Sun Sep 20 17&#8203;:13&#8203;:19 2015\, tonyc wrote&#8203;:
> 
> > Oops\, updated patch\.
> > 
> 
> \+1\.  Smoking in branch smoke\-me/jkeenan/126084\-dofile\.
> 

Tony&#8203;:  Smoke reports \(mostly from your own smokers\) can be found here&#8203;:

http&#8203;://perl\.develop\-help\.com/?b=smoke\-me%2Fjkeenan%2F126084\-dofile

My sense is that there are no new test failures as a result of your patch \-\- just the usual suspects\.

Shall we proceed to apply?

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue\, 29 Sep 2015 22&#8203;:09&#8203;:03 GMT\, jkeenan wrote&#8203;:
> On Mon Sep 21 18&#8203;:28&#8203;:35 2015\, jkeenan wrote&#8203;:
> > On Sun Sep 20 17&#8203;:13&#8203;:19 2015\, tonyc wrote&#8203;:
> >
> > > Oops\, updated patch\.
> > >
> >
> > \+1\.  Smoking in branch smoke\-me/jkeenan/126084\-dofile\.
> >
> 
> Tony&#8203;:  Smoke reports \(mostly from your own smokers\) can be found here&#8203;:
> 
> http&#8203;://perl\.develop\-help\.com/?b=smoke\-me%2Fjkeenan%2F126084\-dofile
> 
> My sense is that there are no new test failures as a result of your
> patch \-\- just the usual suspects\.
> 
> Shall we proceed to apply?
> 
> Thank you very much\.

Though this branch got smoked\, it was never applied to blead and the ticket remains unresolved\.

TonyC\, can you take a look?

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @tonycoz

On Thu\, 20 Sep 2018 16&#8203;:30&#8203;:16 \-0700\, jkeenan wrote&#8203;:
> Though this branch got smoked\, it was never applied to blead and the
> ticket remains unresolved\.
> 
> TonyC\, can you take a look?

This was fixed by 33fe1955034f86b7a9abfc24e0d45a5012030aeb\.

Since it was previously fixed I'm resolving it instead of marking it pending release\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
